File: Programmers/[77486] 다단계 칫솔 판매/Johoseong/77486.py
def solution(enroll, referral, seller, amount):
    N = len(enroll)
    costs = { e: 0 for e in enroll }
    indexs = { enroll[i]: i for i in range(N) }
    indexs['-'] = -1
    costs['-'] = 0

    for a, s in enumerate(seller):
        i = indexs[s] # 현재 직원의 인덱스
        cur = enroll[i]
        parent = ''
        cur_cost = amount[a] * 100
        costs[cur] += cur_cost

        while parent != '-':
            parent = referral[i]

            cost = cur_cost
            parent_cost = int(cost * 0.1)
            cur_cost = cost - int(parent_cost)

            if parent_cost < 1: break # 10%가 1보다 작으면 부모에게 분배 안해도 됨

            costs[cur] -= parent_cost # 내 매출에서 10% 차감
            costs[parent] += parent_cost # 부모는 10% 얻음
            i = indexs[parent]
            cur = enroll[i]
            cur_cost = parent_cost
    
    costs.pop('-')
    return list(costs.values())

File: Programmers/[77486] 다단계 칫솔 판매/woong-jae/77486.js
function solution(enroll, referral, seller, amount) {
  class Node {
      constructor(name) {
          this.name = name;
          this.profit = 0;
          this.parent = null;
      }
  }
  // 트리 생성
  const name_to_index = new Map(enroll.map((name, index) => [name, index]));
  const nodes = enroll.map((name) => new Node(name));
  const root = new Node("-");
  
  referral.forEach((parent, index) => {
      const node = nodes[index];
      if(parent === "-") {
          node.parent = root;
          return;
      }
      const parentNode = nodes[name_to_index.get(parent)];
      node.parent = parentNode;
  });
  
  amount.forEach((sold, index) => {
      let earned = sold * 100;
      let cur = nodes[name_to_index.get(seller[index])];
      while(cur) {
          if(earned === 0) return;
          const will_earn = Math.ceil(earned * 0.9);
          earned -= will_earn;
          
          cur.profit += will_earn;
          cur = cur.parent;
      }
  });
  
  return nodes.map(({ profit }) => profit);
}